### ISSUE TITLE
Overriding find_program on the command line

### ISSUE THREAD

**Issue by ryandesign** [mesonbuild/meson#13096]
I wish to build someone else's software that uses meson. In their meson.build file they have written:

```
  m4 = find_program('m4', required: true)
```

I do not want meson to search for a program called `m4`. Instead, I want to inform meson which program it should use for this purpose (specifically `/opt/local/bin/gm4`). I don't want to patch the meson.build file. If this were an autoconf-generated configure script, it would be a simple matter of specifying an argument like `ac_cv_prog_M4=/opt/local/bin/gm4` on the command line when I run `./configure`. What's the meson equivalent?

I have read [How do I do X in Meson?](https://mesonbuild.com/howtox.html) and the documentation for [find_program()](https://mesonbuild.com/Reference-manual_functions.html#find_program) and [meson.override_find_program()](https://mesonbuild.com/Reference-manual_builtin_meson.html#mesonoverride_find_program) and I haven't found anything that seems applicable to this.

**Comment by thesamesam** [comment]
@eli-schwartz I think I wanted this before for Valgrind, maybe.

**Comment by thesamesam** [comment]
@ryandesign You can do what we do in our zstd packaging:

```meson
        # This replaces the no-find-valgrind patch once bugfix lands in a meson
        # release + we can BDEPEND on it (https://github.com/mesonbuild/meson/pull/11372)
        cat >> ${native_file} <<-EOF || die
        [binaries]
        valgrind='valgrind-falseified'
        EOF

        meson setup ... --native-file "${native_file}"
```

**Comment by eli-schwartz** [comment]
A machine file can easily specify the exact command to use. Specifying this on the command line is not currently possible (but has been discussed).

**Comment by tristan957** [comment]
Related: https://github.com/mesonbuild/meson/pull/12623. You have to convince @jpakkane

**Comment by ryandesign** [comment]
> You can do what we do in our zstd packaging

Thanks. It's less convenient to have to write these options to a separate file than to specify them on the command line directly, but it works—for `find_program`. I'm seeing meson.build files that find python by using [`find_installation`](https://mesonbuild.com/Python-module.html#find_installation) instead, and this method doesn't seem to work for that. How do I override the results of `find_installation`? How many more ways are there that meson.build files typically find programs, and how do I override their results from the command line or from a machine file?

**Comment by ryandesign** [comment]
> this method doesn't seem to work for […] `find_installation`

Scratch that… the meson.build file said:

```python
python = import('python').find_installation('python3')
```

so I assumed what I would have to add to the machine file was:

```ini
[binaries]
python3='/opt/local/bin/python3.12'
```

That didn't work. What did work however was:

```ini
[binaries]
python='/opt/local/bin/python3.12'
```

**Comment by ryandesign** [comment]
I've now convinced myself by looking through the python module's `find_installation` function that the `python` binary key is indeed the correct one to use:

https://github.com/mesonbuild/meson/blob/9e3b3db7054c7dedecd14db3e6061ff7e2227faf/mesonbuild/modules/python.py#L493

However searching through the meson codebase for other instances of `lookup_binary_entry` I found the python3 module and it uses the key `python3`:

https://github.com/mesonbuild/meson/blob/9e3b3db7054c7dedecd14db3e6061ff7e2227faf/mesonbuild/modules/python3.py#L60

So to override a particular build system's python choice, it sounds like I have to read the meson.build file to discover whether it is using the python or python3 module.

All of this information on how to override a meson.build file's program choices would be useful to have in the documentation.

The "How do I do X in Meson?" document seems to be a mix of information targeted at meson.build authors and users building software that uses a meson.build file. It might be useful to separate it clearly into two sections for those two different audiences. The section aimed at users should not assume any prior knowledge of meson or python—most users just want to build a piece of software, not understand its build system.

**Comment by dcbaker** [comment]
The python3 module is deprecated, so I wouldn't worry about what it's doing. I'm not sure that there's even anyone using it anymore, or has been for some time.

**Comment by eli-schwartz** [comment]
I think in the past couple of years I have seen maybe two projects that used it, and both were only using it because they started using it before the "python" module replaced it.

**Comment by bonzini** [comment]
Duplicate of #12623

**Comment by ryandesign** [comment]
Does #12623 also cover updating the documentation in the way I described?

**Comment by tristan957** [comment]
That PR is on hold until other maintainers agree with adding that support.

**Comment by ryandesign** [comment]
Per the proceeding comments in this issue, some of what I want is already possible, I just didn't know because I couldn't find it in the documentation. My question is whether the documentation can be updated to describe the existing functionality.

**Comment by bonzini** [comment]
If you want to give it a shot, you're probably the best person to identify the hole in the documentation and fix it.

**Comment by ryandesign** [comment]
I am not currently working on updating meson documentation. I have pointed out in this issue which documentation I read that did not answer my question. Those would be good places for someone familiar with how the documentation is edited to add information. Perhaps this issue could be reopened until it is resolved by improving the documentation.

**Comment by bonzini** [comment]
The bug is about "Overriding find_program on the command line", and that part is a duplicate of the other bug.

At this point I don't understand exactly what confused you, for example whether the proposed fix is related to m4 as in the description or to python as in the subsequent comments. Therefore it would be helpful if you could sum up what parts of the issue you would like to have improved documentation about.

My original suggestion that you could also adjust the wording based on what you would have liked to find would not be a lot more work than that, but if you prefer not to do that, that's fine. However, let me remind you that you're using an open source project that is maintained by volunteers (as exemplified by the fact that I am writing on an issue tracker on Sunday morning), with no professional tech writers on sight. So even if you're not working on the Meson documentation, it's normal to ask for help because other people *would* actually give it a shot if encouraged.

**Comment by ryandesign** [comment]
> The bug is about "Overriding find_program on the command line" 

I know what the bug is about, because I filed it.

> At this point I don't understand exactly what confused you, for example whether the proposed fix is related to m4 as in the description or to python as in the subsequent comments.

My preceding comments are clear, so if you didn't understand them, read them again. To your specific question of whether I'm interested in m4 or python, those were just two examples. The answer is that I'm interested in all binaries. Whatever binary a meson.build file is trying to find, I want to be able to override it with my choice when I build.

> My original suggestion that you could also adjust the wording based on what you would have liked to find would not be a lot more work than that, but if you prefer not to do that, that's fine. However, let me remind you that you're using an open source project that is maintained by volunteers (as exemplified by the fact that I am writing on an issue tracker on Sunday morning), with no professional tech writers on sight. So even if you're not working on the Meson documentation, it's normal to ask for help because other people *would* actually give it a shot if encouraged.

I know how open-source software works. I've been a manager of the MacPorts project since 2009 and have managed its automated build system since 2016, on top of maintaining hundreds of ports and contributing bug reports and fixes to countless other projects. As such, I have my hands and brain full of other tasks. I do not have room to add learning about your documentation system to my responsibilities. It's not my job to fix every problem in everybody's project. I'm doing you a courtesy by pointing out where you can improve your project, but I don't have time to keep baby-sitting each issue or re-clarify things that are already clear.

**Comment by bonzini** [comment]
> I know what the bug is about, because I filed it.

Yes, and you made it about three things in three different moments. find_program, find_installation, documentation. If you have multiple issues **file** multiple issues. Anyhow, I will file an issue about the documentation myself now.
